README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
estatapi - 政府統計の総合窓口（e-Stat）のAPIを使うためのRパッケージ
===================================================================

[![Travis-CI Build Status](https://travis-ci.org/yutannihilation/estatapi.svg?branch=master)](https://travis-ci.org/yutannihilation/estatapi) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/yutannihilation/estatapi?branch=master&svg=true)](https://ci.appveyor.com/project/yutannihilation/estatapi) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/estatapi)](http://cran.r-project.org/package=estatapi)

*English version of README is [here](https://github.com/yutannihilation/estatapi/blob/master/README.en.md)*

e-Stat APIとは
--------------

e-Statは日本の統計情報が集まっているポータルサイトです。e-StatにはAPIが用意されていて、[ウェブサイト](http://www.e-stat.go.jp/api/)上でアカウントを登録すると使えるようになります。

**このサービスは、政府統計総合窓口(e-Stat)のAPI機能を使用していますが、サービスの内容は国によって保証されたものではありません。**

事前準備
--------

[利用ガイド](http://www.e-stat.go.jp/api/api-guide/)に従ってアプリケーションIDを取得してください。APIにアクセスする際は、`appId`というパラメータに取得したアプリケーションIDを指定します。

インストール
------------

estatapiはCRAN上に公開されているパッケージなので、通常通りインストールできます。

``` r
install.packages("estatapi")
```

開発版をインストールするには`devtools::install_github()`を使ってください。

``` r
devtools::install_github("yutannihilation/estatapi")
```

使い方
------

現在、このパッケージではバージョン2.0のうち4つをサポートしています。各APIの詳しい解説やパラメータの指定の仕方、返ってくる結果の意味は、公式ドキュメントを参照してください。

-   [統計表情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_1): 提供されている統計表を検索します。
-   [メタ情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_2): 統計データのメタ情報を取得します。
-   [統計データ取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_3): 統計データを取得します。
-   [データカタログ情報取得](http://www.e-stat.go.jp/api/e-stat-manual/#api_2_6): 統計表ファイル（Excel、CSV、PDF）および統計データベースの情報を取得します。

### 統計表情報取得（`estat_getStatsList()`）

提供されている統計表を検索します。この関数は、結果を`tbl_df`（dplyrの`data.frame`。`data.frame`とほぼ同じように扱える）として返します。

例えば、「チョコレート」というキーワードを含む統計を検索するときは`searchWord`という引数にキーワードを指定して、以下のようにします。

``` r
appId <- "XXXXXXXXX"
```

``` r
library(estatapi)
#> このサービスは、政府統計総合窓口(e-Stat)のAPI機能を使用していますが、サービスの内容は国によって保証されたものではありません。

estat_getStatsList(appId = appId, searchWord = "チョコレート")
#> # A tibble: 189 x 13
#>           @id        STAT_NAME    GOV_ORG
#>         <chr>            <chr>      <chr>
#> 1  0000100087 全国物価統計調査     総務省
#> 2  0000100104 全国物価統計調査     総務省
#> 3  0000100125 全国物価統計調査     総務省
#> 4  0000100126 全国物価統計調査     総務省
#> 5  0000100127 全国物価統計調査     総務省
#> 6  0000100128 全国物価統計調査     総務省
#> 7  0000100129 全国物価統計調査     総務省
#> 8  0000100130 全国物価統計調査     総務省
#> 9  0003126524     工業統計調査 経済産業省
#> 10 0003143784     工業統計調査 経済産業省
#> # ... with 179 more rows, and 10 more variables: STATISTICS_NAME <chr>,
#> #   TITLE <chr>, CYCLE <chr>, SURVEY_DATE <chr>, OPEN_DATE <chr>,
#> #   SMALL_AREA <chr>, MAIN_CATEGORY <chr>, SUB_CATEGORY <chr>,
#> #   OVERALL_TOTAL_NUMBER <chr>, UPDATED_DATE <chr>
```

ここで、`STAT_NAME`や`GOV_ORG`は人間が読みやすい形式のラベルになっていますが、 プログラム中で扱う場合はコードのままの方が都合がいいこともあります。そのときは`.use_label = FALSE`を指定してください。

``` r
estat_getStatsList(appId = appId, searchWord = "チョコレート", .use_label = FALSE)
#> # A tibble: 189 x 13
#>           @id STAT_NAME GOV_ORG
#>         <chr>     <chr>   <chr>
#> 1  0000100087  00200572   00200
#> 2  0000100104  00200572   00200
#> 3  0000100125  00200572   00200
#> 4  0000100126  00200572   00200
#> 5  0000100127  00200572   00200
#> 6  0000100128  00200572   00200
#> 7  0000100129  00200572   00200
#> 8  0000100130  00200572   00200
#> 9  0003126524  00550010   00550
#> 10 0003143784  00550010   00550
#> # ... with 179 more rows, and 10 more variables: STATISTICS_NAME <chr>,
#> #   TITLE <chr>, CYCLE <chr>, SURVEY_DATE <chr>, OPEN_DATE <chr>,
#> #   SMALL_AREA <chr>, MAIN_CATEGORY <chr>, SUB_CATEGORY <chr>,
#> #   OVERALL_TOTAL_NUMBER <chr>, UPDATED_DATE <chr>
```

### メタ情報取得（`estat_getMetaInfo()`）

統計データのメタ情報を取得します。この関数は、結果を`list`として返します。`list`の各要素が、それぞれのデータ項目についてのメタ情報を含んだ`tbl_df`になっています。

例えば、`0003103532`というIDの統計に関するメタ情報を取得するには、`statsDataId`という引数にIDを指定して、以下のようにします。

``` r
meta_info <- estat_getMetaInfo(appId = appId, statsDataId = "0003103532")

names(meta_info)
#> [1] "tab"    "cat01"  "cat02"  "area"   "time"   ".names"

meta_info$cat01
#> # A tibble: 703 x 5
#>        @code                              @name @level    @unit
#>        <chr>                              <chr>  <chr>    <chr>
#> 1  000100000           世帯数分布（抽出率調整）      1 一万分比
#> 2  000200000                         集計世帯数      1     世帯
#> 3  000300000                           世帯人員      1       人
#> 4  000400000                       18歳未満人員      2       人
#> 5  000500000                       65歳以上人員      2       人
#> 6  000600000                 65歳以上無職者人員      3       人
#> 7  000700000                           有業人員      1       人
#> 8  000800000                       世帯主の年齢      1       歳
#> 9  000900000                             持家率      1       ％
#> 10 001000000 家賃・地代を支払っている世帯の割合      1       ％
#> # ... with 693 more rows, and 1 more variables: @parentCode <chr>
```

### 統計データ取得（`estat_getStatsData()`）

統計データを取得します。この関数は、結果をメタ情報と紐づけて`tbl_df`として返します。

必ず指定しなくてはいけないのは`appId`と`statsDataId`だけですが、それだけだとデータがかなり大きくなって取得に時間がかかります。`cdCat01`（分類事項01）などを指定して必要な項目だけに絞ることをおすすめします。他に絞り込みに指定できるパラメータについては公式ドキュメントを参照してください。

``` r
estat_getStatsData(
  appId = appId,
  statsDataId = "0003103532",
  cdCat01 = c("010800130","010800140")
)
#> Fetching record 1-13184... (total: 13184 records)
#> # A tibble: 13,184 x 12
#>    tab_code 表章項目 cat01_code 品目分類（27年改定） cat02_code
#>       <chr>    <chr>      <chr>                <chr>      <chr>
#> 1        01     金額  010800130     352 チョコレート         03
#> 2        01     金額  010800130     352 チョコレート         03
#> 3        01     金額  010800130     352 チョコレート         03
#> 4        01     金額  010800130     352 チョコレート         03
#> 5        01     金額  010800130     352 チョコレート         03
#> 6        01     金額  010800130     352 チョコレート         03
#> 7        01     金額  010800130     352 チョコレート         03
#> 8        01     金額  010800130     352 チョコレート         03
#> 9        01     金額  010800130     352 チョコレート         03
#> 10       01     金額  010800130     352 チョコレート         03
#> # ... with 13,174 more rows, and 7 more variables: 世帯区分 <chr>,
#> #   area_code <chr>, 地域区分 <chr>, time_code <chr>,
#> #   時間軸（月次） <chr>, unit <chr>, value <dbl>
```

`limit`で取得する最大のレコード数を、`startPosition`で取得を始めるレコードの位置を指定することもできます。とりあえず少しだけ抜き出して見たい場合や、少しずつデータを取ってきたい場合にはこれらのパラメータが便利です。

``` r
estat_getStatsData(
  appId = appId, statsDataId = "0003103532", cdCat01 = c("010800130","010800140"),
  limit = 3
)
#> Fetching record 1-3... (total: 13184 records)
#> # A tibble: 3 x 12
#>   tab_code 表章項目 cat01_code 品目分類（27年改定） cat02_code
#>      <chr>    <chr>      <chr>                <chr>      <chr>
#> 1       01     金額  010800130     352 チョコレート         03
#> 2       01     金額  010800130     352 チョコレート         03
#> 3       01     金額  010800130     352 チョコレート         03
#> # ... with 7 more variables: 世帯区分 <chr>, area_code <chr>,
#> #   地域区分 <chr>, time_code <chr>, 時間軸（月次） <chr>, unit <chr>,
#> #   value <dbl>

estat_getStatsData(
  appId = appId, statsDataId = "0003103532", cdCat01 = c("010800130","010800140"),
  startPosition = 101,
  limit = 3
)
#> Fetching record 101-103... (total: 13184 records)
#> # A tibble: 3 x 12
#>   tab_code 表章項目 cat01_code 品目分類（27年改定） cat02_code
#>      <chr>    <chr>      <chr>                <chr>      <chr>
#> 1       01     金額  010800130     352 チョコレート         03
#> 2       01     金額  010800130     352 チョコレート         03
#> 3       01     金額  010800130     352 チョコレート         03
#> # ... with 7 more variables: 世帯区分 <chr>, area_code <chr>,
#> #   地域区分 <chr>, time_code <chr>, 時間軸（月次） <chr>, unit <chr>,
#> #   value <dbl>
```

### データカタログ情報取得（`estat_getDataCatalog()`）

統計表ファイル（Excel、CSV、PDF）および統計データベースの情報を取得できます。

このAPIはファイルのURLを返すだけなので、そのままRで処理することは難しいかもしれません。

``` r
catalog1 <- estat_getDataCatalog(appId = appId, searchWord = "チョコレート", dataType = c("PDF", "XLS"))

catalog1[1, c("@id", "STAT_NAME", "TABLE_NAME", "SURVEY_DATE", "TABLE_SUB_CATEGORY1", "DATASET_NAME", "NAME", "LANDING_PAGE", "URL", "FORMAT")] %>%
  glimpse
#> Observations: 1
#> Variables: 10
#> $ @id                 <chr> "000000701890"
#> $ STAT_NAME           <chr> "全国物価統計調査"
#> $ TABLE_NAME          <chr> "業態別価格分布－全国，都市階級，都道府県"
#> $ SURVEY_DATE         <chr> "1997"
#> $ TABLE_SUB_CATEGORY1 <chr> "0104チョコレート"
#> $ DATASET_NAME        <chr> "平成9年全国物価統計調査_大規模店舗編_1997年"
#> $ NAME                <chr> "価格分布_1_業態別価格分布－全国，都市階級，都道府県_0104チョコレート"
#> $ LANDING_PAGE        <chr> "http://www.e-stat.go.jp/SG1/estat/GL08020...
#> $ URL                 <chr> "http://www.e-stat.go.jp/SG1/estat/GL08020...
#> $ FORMAT              <chr> "XLS"
```


NEWS.md
# estatapi 0.3.0

## Major Changes

* `estat_getStatsData()` and `estat_getStatsList()` has been overhauled. Accordingly, each function
  returns the result with different colnames than before. I'm sorry, but this change was inevitable
  to fix bugs like [#7](https://github.com/yutannihilation/estatapi/issues/7)... For the details,
  please see [this gist](https://gist.github.com/yutannihilation/e89e3cab6212aa9a390c2c831310ddae).
* Treat most of parameters as formal arguments. Please be careful if you use unnamed arguments.
* Use e-Stat API version 2.1, which is released on Jul 14, 2016.
* `estat_getStatsData()` automatically repeats fetching records when the records are more than the
  number that can be fetched at one time (=100,000 records). You can stop this behaviour by specifying
  `limit` or `.fetch_all` argument.
* Add `getDataCatalog()`, which finds statistical dataset files (Excel, CSV and PDF) and databases.

## Minor Changes

* Show license notice on startup.


tests/testthat/test-api.R
context("API with mocks")

dummy_res <- structure(
  list(
    url = "",
    status_code = 200L,
    headers = structure(
      list(
        `content-type` = "application/json;charset=utf-8",
        `content-encoding` = "gzip"
      ),
      .Names = c("content-type",
                 "content-encoding"),
      class = c("insensitive", "list")
    ),
    content = raw(0)
  ),
  .Names = c("url",
             "status_code",
             "headers",
             "content"),
  class = "response"
)

test_that("estat_getStatsList processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...)
      purrr::update_list(dummy_res, content = readRDS("content_getStatsList.rds")),
    expect_identical(
      estat_getStatsList(
        appId = "XXXX",
        searchWord = "CD",
        limit = 3
      ),
      readRDS("result_getStatsList.rds")
    )
  )
})

test_that("estat_getStatsList (w/o label) processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...)
      purrr::update_list(dummy_res, content = readRDS("content_getStatsList.rds")),
    expect_identical(
      estat_getStatsList(
        appId = "XXXX",
        searchWord = "CD",
        limit = 3,
        .use_label = FALSE
      ),
      readRDS("result_getStatsList_wo_label.rds")
    )
  )
})

test_that("estat_getMetaInfo processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...)
      purrr::update_list(dummy_res, content = readRDS("content_getMetaInfo.rds")),
    expect_identical(
      estat_getMetaInfo(appId = "XXXX", statsDataId = "0003065345"),
      readRDS("result_getMetaInfo.rds")
    )
  )
})

test_that("estat_getStatsData processes the API response as expected", {
  with_mock(
    `estatapi:::estat_getStatsDataCount` = function(...) 40014,
    `estatapi::estat_getMetaInfo` = function(...) readRDS("result_getMetaInfo.rds"),
    `httr::GET` = function(...)
      purrr::update_list(dummy_res,
                         content = readRDS("content_getStatsData.rds")),
    expect_identical(
      estat_getStatsData(
        appId = "XXXX",
        statsDataId = "0003065345",
        cdCat01 = c("008", "009", "010"),
        limit = 3
      ),
      readRDS("result_getStatsData.rds")
    )
  )
})


test_that("estat_getDataCatalog processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...)
      purrr::update_list(dummy_res, content = readRDS("content_getDataCatalog.rds")),
    expect_identical(
      estat_getDataCatalog(
        appId = "XXXX",
        searchWord = "\u30c1\u30e7\u30b3\u30ec\u30fc\u30c8",
        dataType = c("PDF", "XLS")
      ),
      readRDS("result_getDataCatalog.rds")
    )
  )
})


test_that("estat_getDataCatalog (w/o label) processes the API response as expected", {
  with_mock(
    `httr::GET` = function(...)
      purrr::update_list(dummy_res, content = readRDS("content_getDataCatalog.rds")),
    expect_identical(
      estat_getDataCatalog(
        appId = "XXXX",
        searchWord = "\u30c1\u30e7\u30b3\u30ec\u30fc\u30c8",
        dataType = c("PDF", "XLS"),
        .use_label = FALSE
      ),
      readRDS("result_getDataCatalog_wo_label.rds")
    )
  )
})


tests/testthat/test-api-real.R
context("API with real request")


# Constants ---------------------------------------------

EXPECTED_NAMES <- c("tab_code",   "\u8868\u7ae0\u9805\u76ee",
                    "cat01_code", "\u54c1\u76ee\u5206\u985e\uff0827\u5e74\u6539\u5b9a\uff09",
                    "cat02_code", "\u4e16\u5e2f\u533a\u5206",
                    "area_code",  "\u5730\u57df\u533a\u5206",
                    "time_code",  "\u6642\u9593\u8ef8\uff08\u6708\u6b21\uff09",
                    "unit",
                    "value")

STATS_DATA_ID <- "0003103532"

CD_CAT01_SMALL <- c("010800130","010800140")
EXPECTED_COUNT_SMALL <- 13184

CD_CAT01_LARGE <- c("000100000", "000200000", "000300000", "000400000", "000500000",
               "000600000", "000700000", "000800000", "000900000", "001000000",
               "001100000", "010000000", "010100000", "010110001", "010120000",
               "010120010", "010120020")
EXPECTED_COUNT_LARGE <- 110864


# Util Functions ----------------------------------------

wrap_api_func <- function(fun, ...) {
  skip_on_cran()

  if (!file.exists("../../.appId")) {
    skip("appId is not available")
  }

  appId <- readr::read_lines("../../.appId")
  purrr::partial(fun, appId = appId)
}

check_df <- function(d, expected_count, expected_names = EXPECTED_NAMES) {
  # Confirm there are no duplicated rows
  expect_equal(nrow(d), nrow(dplyr::distinct(d)))

  # Confirm the number of rows is the same as expected
  expect_equal(nrow(d), expected_count)

  # Confirm the column names are the same as expected
  expect_equal(names(d), expected_names)
}


# Tests with <1000000 records -----------------------------------------------

test_that("estat_getStatsData with <100000 records works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL
  )

  check_df(d, EXPECTED_COUNT_SMALL)
})

test_that("estat_getStatsData with <100000 records and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    limit = 10
  )

  check_df(d, 10)
})

test_that("estat_getStatsData with <100000 records and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    limit = 10
  )

  check_df(d, 10)
})


test_that("estat_getStatsData with <100000 records and startPosition works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    startPosition = 10001
  )

  check_df(d, EXPECTED_COUNT_SMALL - 10001 + 1)
})

test_that("estat_getStatsData with <100000 records and startPosition and limit works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_SMALL,
    startPosition = 10001,
    limit = 10
  )

  check_df(d, 10)
})


# Tests with >1000000 records -----------------------------------------------

test_that("estat_getStatsData with >100000 records works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE
  )

  check_df(d, EXPECTED_COUNT_LARGE)
})

test_that("estat_getStatsData with >100000 records and limit >100000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    limit = 110000
  )

  check_df(d, 110000)
})

test_that("estat_getStatsData with >100000 records and limit <100000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    limit = 90000
  )

  check_df(d, 90000)
})

test_that("estat_getStatsData with >100000 records and startPosition works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001
  )

  check_df(d, EXPECTED_COUNT_LARGE - 10001 + 1)
})

test_that("estat_getStatsData with >100000 records and startPosition and limit >1000000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001,
    limit = 100800
  )

  check_df(d, 100800)
})

test_that("estat_getStatsData with >100000 records and startPosition and limit <1000000 works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = STATS_DATA_ID,
    cdCat01 = CD_CAT01_LARGE,
    startPosition = 10001,
    limit = 90000
  )

  check_df(d, 90000)
})


# Test English -----------------------------------------------

test_that("estat_getStatsData with <100000 records in English works fine", {
  func_getStatsData <- wrap_api_func(estat_getStatsData)
  d <- func_getStatsData(
    statsDataId = "0003036792",
    cdTab = "2",
    cdArea = "13A01",
    cdCat01 = "7009",
    lang = "E"
  )

  check_df(d, 578, c("tab_code", "Tabulated variable",
                     "cat01_code", "Items",
                     "area_code", "AREA",
                     "time_code", "Time",
                     "unit",
                     "value"))
})



tests/testthat/test-try_dollar.R
context("try_dollar")

test_that("If x is a scalar value, return x as it is", {
  expect_equal(try_dollar(1), 1)
  expect_equal(try_dollar(TRUE), TRUE)
  expect_equal(try_dollar("a"), "a")
  expect_equal(try_dollar(as.Date("2015-01-01")), as.Date("2015-01-01"))
})

test_that("If x is a list with no `$` element, return x as it is", {
  expect_equal(try_dollar(list()), list())
  expect_equal(try_dollar(list(a = 1, b = 2)), list(a = 1, b = 2))
  expect_equal(try_dollar(iris), iris)
})

test_that("If x is a list with a `$` element, return `$` when .use_label = TRUE. Otherwise, return the other one", {
  expect_equal(try_dollar(list("$" = 1, "@id" = 111)), 1)
  expect_equal(try_dollar(list("$" = 1, "@id" = 111), .use_label = FALSE), 111)
})


R/util.R
#' Util

ESTAT_API_URL <- "http://api.e-stat.go.jp/"

#' @title e-Stat API
#'
#' @description Get Statistical Something From e-Stat API
#'
#' @param path API endpoint
#' @param appId application ID
#' @param ... other parameters
#'
#' @export
estat_api <- function(path, appId, ...) {
  # convert params like cdCat01=c("001","002") to cdCat01="001,002"
  query <- flatten_query(list(appId = appId, ...))

  res <- httr::GET(
    ESTAT_API_URL,
    config = httr::add_headers(`Accept-Encoding` = "gzip"),
    path = path,
    query = query
  )

  httr::warn_for_status(res)

  media <- httr::parse_media(res$headers$`content-type`)

  # A result of gettStatsData, getMetaInfo, getDataCatalog and getStatsList is JSON.
  # That of getSimpleStatsData is CSV (but the media type is "text/plain").
  if (media$type == "application" && media$subtype == "json") {
    parse_result_json(res)
  } else if (media$type == "text" && media$subtype == "plain") {
    parse_result_csv(res)
  } else {
    stop("unknown media type: ", res$headers$`content-type`)
  }
}

flatten_query <- function(x) {
  x <- x %>%
    purrr::compact() %>%
    purrr::map(~ paste0(as.character(.), collapse = ","))

  # Ignore duplicated elements
  x[unique(names(x))]
}

as_flattened_character <- function(x, .use_label = TRUE) {
  x %>%
    purrr::map(try_dollar, .use_label = .use_label) %>%
    purrr::flatten() %>%
    purrr::map(as.character)
}

# 1) scalar value  ->  return x as it is
# 2) list with "$" element
#   2-1) if .use_label is TRUE  ->  return the value of "$"
#   2-2) if .use_label is FALSE ->  return the other value
# 3) list without "$" element  ->  return x as it is (x will be flattened outside of this function)
try_dollar <- function(x, .use_label = TRUE) {
  if(! is.list(x)) return(x)
  if(! "$" %in% names(x)) return(x)

  if(.use_label) {
    x[["$"]]
  } else {
    x[[which(names(x) != "$")]]
  }
}

get_class_info <- function(class_obj) {
  meta_ids <- purrr::map_chr(class_obj, "@id")
  meta_names <- purrr::map_chr(class_obj, "@name")

  class_info <- purrr::map(class_obj, ~ dplyr::bind_rows(.$CLASS))
  names(class_info) <- meta_ids

  purrr::update_list(class_info, .names = dplyr::data_frame(id = meta_ids, name = meta_names))
}

merge_class_info <- function(value_df, class_info, name) {
  info <- class_info[[name]] %>%
    dplyr::select_("`@code`", "`@name`")

  key <- sprintf("@%s", name)
  colnames(info) <- c(key, sprintf("%s_info", name))
  dplyr::left_join(value_df, info, by = key)
}

parse_result_json <- function(res) {
  result_parsed <- httr::content(res)

  if (result_parsed[[1]]$RESULT$STATUS != 0) {
    stop(result_parsed[[1]]$RESULT$ERROR_MSG)
  }

  result_parsed
}

# Example responses of getStatsData:
#
# 1) success
#     "VALUE"
#     "tab_code","XXXX","cat01_code","YYYY",...
#
# 2) fail
#     "RESULT"
#     "STATUS","100"
#     "ERROR_MSG","..."
#     "DATE","2016-XX-XXTXX:XX:XX.XXX+09:00"
parse_result_csv <- function(res) {
  result_text <- httr::content(res, as = "text")

  if (readr::read_lines(result_text, n_max = 1) != "\"VALUE\"") {
    stop(result_text)
  }

  readr::read_csv(result_text,
                  skip = 1,
                  col_types = readr::cols(
                    value    = readr::col_number(),
                    .default = readr::col_character()
                  ))
}


R/getMetaInfo.R
#' getMetaInfo API
#'
#' Get meta-information about the statistical dataset via e-Stat API.
#'
#' @param appId Application ID
#' @param statsDataId ID of the statistical dataset
#' @param lang Language of data. \code{"J"}(Japanese) or \code{"E"}(English).
#' @param ... Other parameters.
#'
#' @seealso
#' \url{http://www.e-stat.go.jp/api/e-stat-manual2-1/#api_2_2}
#' \url{http://www.e-stat.go.jp/api/e-stat-manual2-1/#api_3_3}
#'
#' @examples
#' \dontrun{
#' estat_getMetaInfo(appId = "XXXX", statsDataId = "0003065345")
#' }
#'
#' @export
estat_getMetaInfo <- function(appId, statsDataId, lang = c("J", "E"), ...) {
  lang <- match.arg(lang)
  j <- estat_api("rest/2.1/app/json/getMetaInfo", appId = appId, statsDataId = statsDataId, lang = lang, ...)

  get_class_info(j$GET_META_INFO$METADATA_INF$CLASS_INF$CLASS_OBJ)
}


R/getStatsData.R
#' getStatsData API
#'
#' Get some statistical data via e-Stat API.
#'
#' @param appId application ID
#' @param statsDataId ID of the statistical dataset
#' @param startPosition Integer. The the first record to get.
#' @param limit Integer. Max number of records to get.
#' @param lang Language of data. \code{"J"}(Japanese) or \code{"E"}(English).
#' @param .fetch_all Whether to fetch all records when the number of records
#'                    is larger than 100,000.
#' @param ... Other parameters like \code{lvCat01} and \code{cdCat01}.
#'    See \code{Other parameters} section for more details.
#'
#' @seealso
#' \url{http://www.e-stat.go.jp/api/e-stat-manual2-1/#api_2_3}
#' \url{http://www.e-stat.go.jp/api/e-stat-manual2-1/#api_3_4}
#' @section Other parameters:
#' For every detailed information, please visit the URL in See Also.
#' \itemize{
#'  \item \code{lvTab}:
#'    Level of the meta-information. The format can be \code{X} or \code{X-Y}, \code{-X} and \code{X-}.
#'  \item \code{cdTab}:
#'    Code(s) of the meta-infomation items to select. The format can be a character vector (\code{c("001", "002")}) or
#'    a chraracter of values and commas (\code{"001,002"}).
#'  \item \code{cdTabFrom}:
#'    The code of the first meta-information item to select.
#'  \item \code{cdTabTo}:
#'    The code of the last meta-information item to select.
#'  \item \code{lvTime}:
#'     Level of the time to select. The format is the same as \code{lvTab}
#'  \item \code{cdTime}
#'     Time(s) to select. The format is the same way like \code{cdTab}
#'  \item \code{cdTimeFrom}:
#'    The first time to select. The format is the same way like \code{cdTabFrom}
#'  \item \code{cdTimeTo}:
#'    The last time to select. The format is the same way like \code{cdTabTo}
#'  \item \code{lvArea}:
#'     Level of the area to select. The format is the same as \code{lvTab}
#'  \item \code{cdArea}
#'     Code(s) of the Area to select. The format is the same way like \code{cdTab}
#'  \item \code{cdAreaFrom}:
#'    The code of the first area to select. The format is the same way like \code{cdTabFrom}
#'  \item \code{cdAreaTo}:
#'    The code of the last area to select. The format is the same way like \code{cdTabTo}
#'  \item \code{lvCat01}, \code{cdCat01}, \code{cdCat01From}, \code{cdCat01To}, ...:
#'    The same way like above.
#' }
#'
#' @examples
#' \dontrun{
#' # fetch all data, which may take time
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345"
#' )
#'
#' # fetch data up to 10 records
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   limit = 10
#' )
#'
#' # fetch data up to 100,000 records (max number of records available at once)
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   .fetch_all = FALSE
#' )
#'
#' # fetch all data in the specifed category
#' estat_getStatsData(
#'   appId = "XXXX",
#'   statsDataId = "0003065345",
#'   cdCat01 = c("008", "009", "010")
#' )
#'
#' }
#'
#' @export
estat_getStatsData <- function(appId, statsDataId,
                               startPosition = NULL,
                               limit = NULL,
                               lang = c("J", "E"),
                               .fetch_all = TRUE,
                               ...) {
  lang <- match.arg(lang)

  meta_info <- estat_getMetaInfo(appId, statsDataId, lang = lang)
  total_record_count <- estat_getStatsDataCount(appId, statsDataId, lang = lang, ...)
  ranges <- calc_ranges(startPosition, limit, total_record_count, .fetch_all)

  result <- list()

  for (i in seq_len(nrow(ranges))) {
    cur_limit <- ranges$limits[i]
    cur_start <- ranges$starts[i]

    message(sprintf("Fetching record %.0f-%.0f... (total: %.0f records)\n",
                    cur_start, cur_start + cur_limit - 1, total_record_count))

    j <- estat_api("rest/2.1/app/json/getStatsData",
                             appId = appId,
                             statsDataId = statsDataId,
                             startPosition = format(cur_start, scientific = FALSE),
                             limit = format(cur_limit, scientific = FALSE),
                             lang = lang,
                             metaGetFlg = "N",
                             ...)

    value_df <- j$GET_STATS_DATA$STATISTICAL_DATA$DATA_INF$VALUE %>%
      dplyr::bind_rows()

    result[[i]] <- value_df
  }

  result_df <- result %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(`$` = suppressWarnings(readr::parse_number(`$`)))

  colnames(result_df) <- colnames(result_df) %>%
    purrr::map_chr(~ paste0(substring(., 2), "_code")) %>%
    dplyr::recode("_code" = "value", "unit_code" = "unit")

  for (meta in purrr::transpose(meta_info$.names)) {
    meta_df <- meta_info[[meta$id]]
    meta_vec <- purrr::set_names(x  = meta_df[["@name"]],
                                 nm = meta_df[["@code"]])
    result_df[, meta$name] <- meta_vec[result_df[[paste0(meta$id, "_code")]]]
  }

  # FIXME: This is a temporal workaround until getSimpleStatsData API is ready.
  # reorder result as the same order as getSimpleStatsData
  colnames_all <- colnames(result_df)
  colnames_meta <- meta_info$.names$name
  colnames_orig <- dplyr::setdiff(colnames_all, colnames_meta)

  suppressWarnings({
    sorted_colnames <- unlist(purrr::transpose(
        list(colnames_orig, colnames_meta)
      ))
  })
  result_df[, sorted_colnames]
}


estat_getStatsDataCount <- function(appId, statsDataId, ...) {
  j <- estat_api("rest/2.1/app/json/getStatsData",
                 appId = appId,
                 statsDataId = statsDataId,
                 metaGetFlg = "N",
                 cntGetFlg = "Y",
                 ...)

  as.numeric(j$GET_STATS_DATA$STATISTICAL_DATA$RESULT_INF$TOTAL_NUMBER)
}


calc_ranges <- function(startPosition,
                        limit,
                        record_count,
                        .fetch_all,
                        .max_records_at_once = 100000) {
  ranges <- list()

  if (is.null(startPosition)) {
    startPosition <- 1
  }

  if (is.null(limit)) {
    endPosition <- record_count
  } else {
    endPosition <- min(startPosition + limit - 1, record_count)
  }

  if (.fetch_all) {
    ranges$starts <- seq(from = startPosition, to = endPosition, by = .max_records_at_once)
    ranges$limits <- rep(.max_records_at_once, length(ranges$starts))
    # treat a fraction
    fraction <- (endPosition - startPosition + 1) %% .max_records_at_once
    if (fraction != 0) {
      ranges$limits[length(ranges$limits)] <- fraction
    }
  } else {
    ranges$starts <- startPosition
    ranges$limits <- min(startPosition + .max_records_at_once - 1, endPosition)
  }

  dplyr::as_data_frame(ranges)
}

# cdTab may be (1) NULL, (2) a vector of strings or (3) a string of numbers separated by ","
get_tabs <- function(appId, statsDataId, cdTab) {
  if (is.null(cdTab)) {
    # when (1), try to get all tab from getMetaInfo API
    meta_info <- estat_getMetaInfo(appId = appId, statsDataId = statsDataId)

    # tab does not always exist.
    if (!is.null(meta_info$tab)) {
      meta_info$tab$`@code`
    } else {
      NULL
    }
  } else {
    # when (2) or (3), split it by ","
    unlist(strsplit(cdTab, ",", fixed = TRUE))
  }
}
